public/js/libs/area.chart.js

```diff
@@ -0,0 +1,427 @@
+            .attr("class", `${this.id}Grid`)
+            .call(d3.axisLeft(this.yScale)
+                .tickSize(-this.width)
+                .tickFormat("")
+            )
+
+        // Add the X axis gridlines
+        this.plot.append("g")
+            .attr("class", `${this.id}XGrid`)		
+            .attr("transform", "translate(0," + this.height + ")")
+            .call(d3.axisBottom(this.xScale)
+                .tickSize(-this.height)
+                .tickFormat("")
+            )
+    }
+
+    drawFocus() {    
+        // Create focus object
+        this.focus = this.plot.append("g")
+            .attr("class", `${this.id}focus`)
+
+        // Append circle on the line path
+        this.focus.append("circle")
+            .attr("r", 3.5)
+
+        // Add background rectangle behind the text tooltip
+        this.focus.append("rect")
+            .attr("x", -30)
+            .attr("y", "-32px")
+            .attr("rx", 6)
+            .attr("ry", 6)
+            .attr("width", 220)
+            .attr("height", 20);
+
+        // Add text annotation for tooltip
+        this.focus.append("text")
+            .attr("x", -20)
+            .attr("dy", "-18px")
+            .attr("font-size", "12px")
+            .style("fill", "#0E86D4");
+
+        let vis = this;
+        this.plot
+            .on("mouseover", () => this.focus.style("display", null))
+            .on("mouseout", () => this.focus.style("display", "none"))
+            .on("mousemove", function() {
+                vis.tipMove(this)
+            });
+
+        // Select focus objects and set opacity
+        d3.selectAll(`.${this.id}focus`)
+            .style("opacity", 0.9);
+
+        // Select the circle and style it
+        d3.selectAll(`.${this.id}focus circle`)
+            .style("fill", this.color)
+            .style("opacity", 0)
+
+        // Select the rect and style it
+        d3.selectAll(`.${this.id}focus rect`)
+            .style("fill", "whitesmoke")
+            .style("opacity", 0)
+
+    }
+
+    // Function that adds tooltip on hover
+    tipMove(sel) {
+        // Below code finds the date by bisecting and
+        // Stores the x and y coordinate as variables
+        let x0 = this.xScale.invert(d3.mouse(sel)[0]);
+        // This will select the closest date on the x axiswhen a user hover over the chart
+        let bisectDate = d3.bisector(function(d) {return d.datetime;}).left;
+        let i = bisectDate(this.data, x0, 1);
+        let d0 = this.data[i - 1];
+        let d1 = this.data[i];
+        let d = x0 - d0.datetime > d1.datetime - x0 ? d1 : d0;
+
+        // Place the focus objects on the same path as the line
+        this.focus.attr("transform", `translate(${this.xScale(d.datetime)}, ${this.yScale(d.value)})`); 
+
+        let tooltipText = `${d3.timeFormat("%Y-%m-%d (%H:%M)")(d.datetime)} 
+        Value: ${d3.formatPrefix(".2", d.value)(d.value)}`;
+
+        if (this.feature == "system-metrics") {
```

Comment:
It would be cleaner if you could pass these pieces of configuration in directly rather than switching on the feature/id. That's so that if we need another area chart in future, we don't need to edit various pieces of hard-coded options in the area chart implementation, but instead just pass in the new configuration.

Reply:
I find it hard to pass the configuration indirectly. I had this in mind that If we needed another area chart one would not pass the feature setting of which you get an area chart without specific features that system-metrics charts have e.g outage lines.

Reply:
Updated thanks.